### ISSUE TITLE
[Feature Request] Support hostname canonicalization

### ISSUE THREAD

**Issue by matthew-e-brown** [cyjake/ssh-config#65]
The [`ssh_config(5)` manual page](https://man.openbsd.org/ssh_config) discusses several options related to canonicalizing hostnames. The most fundamental of them is `CanonicalizeHostname`, which is described as follows:

> https://man.openbsd.org/ssh_config#CanonicalizeHostname
>
> Controls whether explicit hostname canonicalization is performed. The default, **no**, is not to perform any name rewriting and let the system resolver handle all hostname lookups. If set to **yes** then, for connections that do not use a **ProxyCommand** or **ProxyJump**, [ssh(1)](https://man.openbsd.org/ssh.1) will attempt to canonicalize the hostname specified on the command line using the **CanonicalDomains** suffixes and **CanonicalizePermittedCNAMEs** rules. If **CanonicalizeHostname** is set to **always**, then canonicalization is applied to proxied connections too.
> 
> If this option is enabled, then the configuration files are processed again using the new target name to pick up any new configuration in matching **Host** and **Match** stanzas. A value of **none** disables the use of a **ProxyJump** host.

As an example, my university has a server called Loki. If I were to use the following configuration:

```ssh-config
Host loki.trentu.ca
        ServerAliveInterval 60
        ServerAliveCountMax 2

Host loki
        User matthew
        CanonicalizeHostname yes
        CanonicalDomains yorku.ca carleton.ca trentu.ca
```

Then running `ssh loki` would have SSH performing the following:

1. Match with `Host loki`.
   1. `User` ← `matthew`
   2. `CanonicalizeHostname` ← `yes`
   3. `CanonicalDomains` ← `yorku.ca carleton.ca trentu.ca`
2. Since `CanonicalizeHostname` is enabled, attempt to find a canonical host:
   1. Attempt to canonicalize with `loki.yorku.ca`; fails.
   2. Attempt to canonicalize with `loki.carleton.ca`; fails.
   3. Attempt to canonicalize with `loki.trentu.ca`; succeeds.
3. Re-parse the SSH configuration file as if I'd typed `ssh loki.trentu.ca`.
4. Match with `Host loki.trentu.ca`
   1. `ServerAliveInterval` ← `60`
   2. `ServerAliveCountMax` ← `2`

Which should give our final configuration as:

```
User matthew
HostName loki.trentu.ca
CanonicalizeHostname yes
CanonicalDomains yorku.ca carleton.ca trentu.ca
ServerAliveInterval 60
ServerAliveCountMax 2
```

If everything goes well, both `SSHConfig.parse(...).compute('loki')` and `ssh -G loki` should spit out the same things for these options. Sadly, this is not the case, as you can see below (`grep` to stop `ssh -G` from spitting out all 100+ defaults):

![ssh-config's 'compile' method only outputs the properties placed directly on the 'loki' host, whereas OpenSSH functions exactly as outlined above.](https://github.com/cyjake/ssh-config/assets/43387454/09495a17-f9c8-4a0f-b88d-0b4d2f9fd9bf)

It may be worth noting that this does _not_ happen recursively; I can't have `loki` re-parse to find `loki.trentu.ca` to then re-parse and find `loki.trentu.ca.othersite.co.uk`, since `loki.trentu.ca` is already canonical and there is no need to run it again.

---

I am more than willing to be the one to take this on as a PR. It seems like it should be trivial to implement, at least for `compute`. I am not sure how or if it even should affect the other functions such as `find`, `append`, and `prepend`. But, again, after some discussion, I'm happy to implement this feature.

Cheers.

[man-page]: https://man.openbsd.org/ssh_config
[sec-domains]: https://man.openbsd.org/ssh_config#CanonicalDomains
[sec-hostname]: https://man.openbsd.org/ssh_config#CanonicalizeHostname

**Comment by cyjake** [comment]
Thanks for this feature request Matthew, very comprehensive writing! I think this needs to be a new api such as `.canonicalize` because it introduces actual hostname handshaking to finalize the ssh config. `.compute` on the other hand, will not be able to carry out this task because it is synchronous. 

The canonicalization process is a bit out of the original scope of this library, as it just parses ssh config text and provide several api to quickly compute the result config, only text to AST and AST to object. But I think it is ok to add it, since it seems no third party libraries is need, will be very happy to receive a pull request about this feature.

**Comment by matthew-e-brown** [comment]
Awesome! I'll start tinkering. I actually didn't think about just making it a separate `.canonicalize` method; I was thinking of making it an overload on `compute` that would take `canonicalize` as a separate boolean parameter to conditionally return a promise. That'd work as far as not changing the public-facing API goes, but it would probably be clunkier to implement. A separate method could just call `this.compute`.

**Comment by cyjake** [comment]
Second thought about the new api, it looks like I've already added some child process calls while implementing `exec` criteria of Match. I think it's ok to add another while implementing `canonical` criteria.

**Comment by matthew-e-brown** [comment]
What do you think about doing it with an overload? It could be made a part of `compute`, but without introducing a breaking change for previous consumers, and also without introducing a whole new API.

```ts
type MaybePromise<T> = T | Promise<T>;

class SSHConfig extends Array<Line> {
  // ...

  // with just the one param, or with false passed, returns the record as normal
  public compute(host: string, canonicalize?: false): Record<string, string | string[]>;
  public compute(options: MatchOptions, canonicalize?: false): Record<string, string | string[]>;

  // with true passed as the second argument, it now returns a promise
  public compute(host: string, canonicalize: true): Promise<Record<string, string | string[]>>;
  public compute(options: string, canonicalize: true): Promise<Record<string, string | string[]>>;

  // implementation signature stays hidden from consumers, but returns record | promise<record>
  public compute(params: string | MatchOptions, canonicalize = false): MaybePromse<Record<string, string | string[]>> {
    if (canonicalize !== true) {
      // ...
      // method as it is
      // ...
    } else {
      return new Promise((resolve, reject) => {
        // pseudocode
        const firstPass = this.compute(params, false);
        if (needsCanonicalize(firstPass)) actualHost = canonicalize(firstPass.Hostname);
        const secondPass = this.compute(actualHost, false);
        resolve({ ...secondPass, ...firstPass });
      });
    }
  }

  // ...
}
```

The only way this would be a breaking change is if some consumers of the library were passing an unused `true` into `compute`.

**Comment by cyjake** [comment]
agree

**Comment by BenjaminGrayNp1** [comment]
I'm not sure an explicit `canonicalize` argument is best. It should at least be an object to allow future options to be set backwards compatibly.

But is there a use case for such fine grained control? I'd expect most users of `.compute` to be expecting whatever SSH would compute the config to be. The interesting choice for this library is whether to be purely syntactical, or perform execs and lookups and whatnot.

For implementation, the specific logic in OpenSSH puts it after the first pass, and triggers the second pass

https://github.com/openssh/openssh-portable/blob/e535fbe2af893046c28adfcd787c1fdbae36a24a/ssh.c#L1189-L1230

This should be a little easier to model now that we support two passes

https://github.com/cyjake/ssh-config/blob/6cde6d3fe81753ae80694caae35b279ad025c78a/src/ssh-config.ts#L208-L212

**Comment by cyjake** [comment]
yep, if it can be implemented with implicit multiple pass, we can keep using `.compute()` without extra flag

**Comment by matthew-e-brown** [comment]
The only problem with doing it implicitly is that it would be impossible to determine the return type statically (for TypeScript).

Canonicalizing a hostname requires a DNS lookup, which is async and means we need to return a `Promise<Record>`. However, if hostname canonicalization doesn't need to happen, we could simply return the `Record` directly. That means that, if desired, consumers who don't need canonicalization don't need to deal with Promises or async code at all.

But if we're going to do the canonicalization implicitly (like OpenSSH does[^1]), making it impossible to denote with TypeScript whether or not it'll return a `Promise`, then I think really there are two options:

- Change `.compute()` to _always_ return `Promise<Record>`; or
- Change `.compute()` to return `Record | Promise<Record>`, and leave it up to the consumer to figure out if they need to `await` or not.

The former makes more sense to me, since it just denotes that `.compute()` is an asynchronous operation in general. The downside is that the API would no longer be backwards compatible. If you're okay with changing the API to always return a Promise, then I support the implicit pass and converting `.compute()` to async. Otherwise, that was the rationale behind the explicit/opt-in flag: in that case, the old `.compute(string) -> Record` signature remains intact for those that don't want to deal with async.


[^1]: They're in C with access to syscalls and stuff like `getaddrinfo` so their version has no `async` to worry about&mdash;so we can't follow their example.

**Comment by BenjaminGrayNp1** [comment]
There is already the `spawnSync` for an `exec` match, which can take an arbitrary amount of time. Wouldn't it be fine to do the canoncialisation synchronously as well?

I think just adding an async variant of `compute` (or making `compute` async in a breaking change, with the old method renamed to `computeSync`) would be the better way to introduce it, without associating async-ness with a particular argument or SSH config field.

**Comment by matthew-e-brown** [comment]
Yes, I think it would be totally fine to do canonicalization synchronously as well. But... is it even possible to do a DNS lookup synchronously with JavaScript? Node provides [one with a callback][dns] and [one that returns a promise][dns-promise] in the DNS module.

The documentation for `dns.lookup` say that it calls `getaddrinfo(3)` directly, makes it seem like our best bet for accurately porting over OpenSSH's functionality. But after a quick skim, I can't find a mention in the documentation for the DNS module for how perform the lookup synchronously.

Aside from that, I like the idea of `compute()` and `computeSync()`. It matches Node's naming conventions.


[dns]: https://nodejs.org/api/dns.html#dnslookuphostname-options-callback
[dns-promise]: https://nodejs.org/api/dns.html#dnspromiseslookuphostname-options

**Comment by BenjaminGrayNp1** [comment]
Oh sorry, I didn't realise there wasn't a sync variant of the lookup. Yeah, that makes things hard. If there's no way to do it, could it just be a documented limitation of the `computeSync` variant if we go that path?

**Comment by matthew-e-brown** [comment]
I just did some Googling, since Node says it just uses libuv to convert the synchronous `getaddrinfo` into an async interface... so surely there's some way to de-async it. [Which... there is!][deasync] ...but, it needs to use some C++ code to interact with Node-API. I think anything that adds `node-gyp` or a C++ compiler as a dependency to an SSH config library is serious overkill.

So, yeah, I think just documenting that `computeSync` doesn't support the hostname stuff (or any other `async` stuff in the future) is the best we could do 😄


[deasync]: https://www.npmjs.com/package/deasync

**Comment by cyjake** [comment]
We can `spawnSync('nslookup', [hostname])` then handle the returned output buffer accordingly. In general, I think a major version bump for the compute/computSync duo with `dns.lookup()` as default is worthy.

**Comment by matthew-e-brown** [comment]
That's perfect! I figured we'd have to use `dns.lookup()` instead of spawning a secondary process because `dig` (what I usually use on \*nix) isn't available on Windows. But it looks like `nslookup` is available everywhere. I had never heard of it before.

**Comment by cyjake** [comment]
@matthew-e-brown @BenjaminGrayNp1 are you guys working on this? If not I think  I can have this feature implemented this week.

**Comment by matthew-e-brown** [comment]
I keep meaning to get into it on the weekends but with my next semester quickly approaching I haven't found the time. Go for it 😁
